### PR TITLE
Ed25519 appendices

### PR DESCRIPTION
**Pull request by olorin**
A few small extras on top of #35 ([diff](https://github.com/ambiata/tinfoil/compare/topic/gadt...topic/salt)).

 - GHC 8 support.
 - ed25519 keygen/sign/verify benchmarks.
 - Clarify general thread-safety status in `tinfoil` - everything is thread-safe unless explicitly marked otherwise.

@markhibberd 

/cc @erikd-ambiata @thumphries

### CODE REVIEW

File: test/bench.hs [hunk 1]
```diff
@@ -101,5 +115,11 @@ main = tinfoilBench [
                              ]
   , env ((,) <$> generate' arbitrary <*> generate' arbitrary) $ \ ~(sk, bs) ->
       bgroup "mac/hmacSHA256" $ [ bench "hmacSHA256" $ nf (hmacSHA256 sk) bs
-                             ]
+                                ]
+  , env genEd25519 $ \ ~(sk, pk, sig, msg) ->
+      bgroup "signing/ed25519" $ [
```
Comment (erikd-ambiata) [inline test/bench.hs]: Are you using `StrictData` (the `~` character) here or something? Is that going to compile with ghc-7.8?

Comment (olorin) [inline test/bench.hs]: Yeah, just a habit of mine with benchmarks. It has always compiled with 7.8 in the past.

**Comment by olorin** [comment]
```
∃ ./dist/build/bench/bench sign
benchmarking signing/ed25519/genKeyPair
time                 47.00 μs   (46.44 μs .. 47.66 μs)
                     0.999 R²   (0.998 R² .. 1.000 R²)
mean                 46.72 μs   (46.56 μs .. 47.01 μs)
std dev              691.8 ns   (378.5 ns .. 1.195 μs)

benchmarking signing/ed25519/signMessage
time                 45.21 μs   (45.17 μs .. 45.25 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 45.19 μs   (45.16 μs .. 45.25 μs)
std dev              133.7 ns   (88.05 ns .. 202.3 ns)

benchmarking signing/ed25519/verifyMessage
time                 120.6 μs   (120.4 μs .. 120.7 μs)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 120.5 μs   (120.4 μs .. 120.6 μs)
std dev              282.4 ns   (220.9 ns .. 353.6 ns)
```

File: test/bench.hs (hunk 1 above)
Comment (erikd-ambiata) [inline test/bench.hs]: TIL! Thanks.

**Comment by erikd-ambiata** [comment]
I only reviewed from 7482edc onwards as I assumed the earlier stuff hasn't changed since I reviewed it before.

:zap:

**Comment by thumphries** [comment]
🔨

**Pull request merged**